cstore_fdw.c

```diff
@@ -733,46 +719,58 @@ CStoreProcessAlterTableCommand(AlterTableStmt *alterStatement)
 }
 
 
-/*
- * DropppedCStoreFilenameList extracts and returns the list of cstore file names
- * from DROP table statement
- */
+/* CStoreTableList returns oids of cstore_fdw tables stored in pg_cstore_tables */
 static List *
-DroppedCStoreFilenameList(DropStmt *dropStatement)
+CStoreTableList(void)
 {
-	List *droppedCStoreFileList = NIL;
+	List *cstoreRelationIdList = NIL;
+	Oid cstoreTablesOid = InvalidOid;
+	Relation heapRelation = NULL;
+	HeapScanDesc scanDesc = NULL;
+	HeapTuple heapTuple = NULL;
+	TupleDesc tupleDescriptor = NULL;
 
-	if (dropStatement->removeType == OBJECT_FOREIGN_TABLE)
+	cstoreTablesOid = get_relname_relid(CSTORE_TABLES_NAME, PG_CATALOG_NAMESPACE);
+	if (cstoreTablesOid == InvalidOid)
 	{
-		ListCell *dropObjectCell = NULL;
-		foreach(dropObjectCell, dropStatement->objects)
-		{
-			List *tableNameList = (List *) lfirst(dropObjectCell);
-			RangeVar *rangeVar = makeRangeVarFromNameList(tableNameList);
+		/* the pg_cstore_tables table does not exist */
+		return NIL;
+	}
 
-			Oid relationId = RangeVarGetRelid(rangeVar, AccessShareLock, true);
-			if (CStoreTable(relationId))
-			{
-				CStoreFdwOptions *cstoreFdwOptions = CStoreGetOptions(relationId);
-				char *defaultfilename = CStoreDefaultFilePath(relationId);
+	heapRelation = heap_open(cstoreTablesOid, AccessShareLock);
+	tupleDescriptor = RelationGetDescr(heapRelation);
 
-				/*
-				 * Skip files that are placed in default location, they are handled
-				 * by sql drop trigger. Both paths are generated by code, use
-				 * of strcmp is safe here.
-				 */
-				if (strcmp(defaultfilename, cstoreFdwOptions->filename) == 0)
-				{
-					continue;
-				}
+	scanDesc = heap_beginscan(heapRelation, SnapshotSelf, 0, NULL);
 
-				droppedCStoreFileList = lappend(droppedCStoreFileList,
-												cstoreFdwOptions->filename);
-			}
-		}
+	heapTuple = heap_getnext(scanDesc, ForwardScanDirection);
+	while (HeapTupleIsValid(heapTuple))
+	{
+		bool isNull = false;
+		Datum logicalrelidDatum = heap_getattr(heapTuple, 1, tupleDescriptor, &isNull);
+		Oid relationId = DatumGetObjectId(logicalrelidDatum);
+
+		cstoreRelationIdList = lappend_oid(cstoreRelationIdList, relationId);
+
+		heapTuple = heap_getnext(scanDesc, ForwardScanDirection);
 	}
 
-	return droppedCStoreFileList;
+	heap_endscan(scanDesc);
+	heap_close(heapRelation, AccessShareLock);
+
+	return cstoreRelationIdList;
+}
+
+
+ /* RemoveRelationStorage marks storage nodes for a relation for deletion */
+static void
+RemoveRelationStorage(Oid relationId)
+{
+	RelFileNode relFileNode = {MyDatabaseTableSpace, MyDatabaseId, relationId};
+	Relation relation = palloc0(sizeof(RelationData));
```

Comment:
When this was called from drop trigger, relation_open() fails because the relation was already dropped. This is to create a  **fake** relation which only has rd_node and rd_backend fields that are sufficient for storage removal.